File: HPWorkpath_v1.6.3/Samples/ExampleAPIServices/source/ConfigSample/src/com/hp/workpath/sample/configsample/task/ConfigReaderTask.java
// Copyright 2018 HP Development Company, L.P.
// SPDX-License-Identifier: MIT
package com.hp.workpath.sample.configsample.task;

import android.os.Handler;
import android.os.Looper;
import android.view.View;

import com.hp.workpath.api.Result;
import com.hp.workpath.api.config.ConfigService;
import com.hp.workpath.sample.configsample.Logger;
import com.hp.workpath.sample.configsample.MainActivity;
import com.hp.workpath.sample.configsample.R;

import org.json.JSONObject;

import java.lang.ref.WeakReference;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

/**
 * Worker task for configuration request
 */
public class ConfigReaderTask {

    private final WeakReference<MainActivity> mContextRef;
    private Result mResult;
    private boolean isUpdate;
    private Throwable mThrowable = null;

    private ExecutorService executor = Executors.newSingleThreadExecutor();
    private Handler handler = new Handler(Looper.getMainLooper());

    public ConfigReaderTask(final MainActivity context, boolean isUpdate) {
        this.mContextRef = new WeakReference<>(context);
        this.mResult = new Result();
        this.isUpdate = isUpdate;
    }

    public void taskExecute() {
        try {
            executor.execute(new Runnable() {
                @Override
                public void run() {
                    try {
                        // call ConfigService API to retrieve configuration for the application
                        onPostExecute(ConfigService.getDefaultConfig(mContextRef.get().getApplicationContext(), mResult));
                    } catch (Throwable t) {
                        mThrowable = t;
                        executor.shutdown();
                        onPostExecute(null);
                    }
                }
            });
        } catch (Exception e) {
            mThrowable = e;
            onPostExecute(null);
            executor.shutdown();
        }
    }

    private void onPostExecute(JSONObject jsonObject) {
        handler.post(new Runnable() {
            @Override
            public void run() {
                mContextRef.get().showProgress(View.GONE);
                if (jsonObject != null && mResult.getCode() == Result.RESULT_OK) {
                    Logger.showResult(mContextRef.get(), mContextRef.get().getString(R.string.success));
                    mContextRef.get().getConfigComplete(jsonObject);
                    if (isUpdate) {
                        mContextRef.get().updatePrintOption(jsonObject);
                    }
                } else if (mThrowable != null) {
                    Logger.showResult(mContextRef.get(), "ConfigService.getDefaultConfig " + mThrowable.getMessage());
                } else {
                    Logger.showResult(mContextRef.get(), "ConfigService.getDefaultConfig", mResult);
                }
            }
        });
    }
}

File: HPWorkpath_v1.6.3/Samples_Kotlin/ExampleAPIServices/source/ConfigSample/src/com/hp/workpath/sample/configsample/task/ConfigReaderTask.kt
// Copyright 2018 HP Development Company, L.P.
// SPDX-License-Identifier: MIT
package com.hp.workpath.sample.configsample.task

import android.view.View
import com.hp.workpath.api.Result
import com.hp.workpath.api.config.ConfigService
import com.hp.workpath.sample.configsample.Logger
import com.hp.workpath.sample.configsample.MainActivity
import com.hp.workpath.sample.configsample.R
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.lang.ref.WeakReference

/**
 * Worker task for configuration request
 */
class ConfigReaderTask(context: MainActivity, isUpdate: Boolean) {

    private val mContextRef: WeakReference<MainActivity> = WeakReference(context)
    private var mThrowable: Throwable? = null
    private val result: Result = Result()
    private val isUpdate: Boolean = isUpdate

    suspend fun execute() {
        var configuration: JSONObject? = null
        // call ConfigService API to retrieve configuration for the application
        try {
            mContextRef.get()?.run {
                configuration = ConfigService.getDefaultConfig(this, result)
            }
        } catch (t: Throwable) {
            mThrowable = t
        }
        onPostExecute(configuration)
    }

    private suspend fun onPostExecute(jsonObject: JSONObject?) {
        withContext(Dispatchers.Main) {
            mContextRef.get()?.run {
                showProgressBar(View.GONE)
                if (jsonObject != null && result.code == Result.RESULT_OK) {
                    Logger.showResult(this, getString(R.string.success))
                    getConfigComplete(jsonObject)
                    if (isUpdate) {
                        updatePrintOption(jsonObject)
                    }
                } else if (mThrowable != null) {
                    Logger.showResult(this, "ConfigService.getDefaultConfig ${mThrowable?.message}")
                } else {
                    Logger.showResult(this, "ConfigService.getDefaultConfig", result)
                }
            }
        }
    }
}